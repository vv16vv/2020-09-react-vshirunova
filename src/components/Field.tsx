import React from "react";
import {Cell} from "./Cell";
import './../css/styles.css';

interface FieldProp {
    width: number;
    height: number;
    filledCells: Array<{ x: number, y: number }>;
}

export const Field: React.FC<FieldProp> = (({width, height, filledCells}) => {
    let divs = [];
    for (let h = 1; h <= height; h++) {
        let cells = [];
        for (let w = 1; w <= width; w++) {
            const isFilled = filledCells.filter(point => point.x === w && point.y === h).length > 0;
            cells.push({
                x: w,
                y: h,
                filled: isFilled
            });
        }
        divs.push(cells);
    }
    return <table>
        {divs.map((line, i) => {
            return <tr>
                {line.map((cell, j) => {
                    return <Cell key={i * j} x={cell.x} y={cell.y} filled={cell.filled}/>
                })}
            </tr>;
        })}
    </table>;
});