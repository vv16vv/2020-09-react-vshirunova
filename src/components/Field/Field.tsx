import React from "react";
import {Cell} from "cmp/Cell";
import {StyledTable, StyledTr} from "styled/StyledComponents";

interface FieldProp {
    width: number;
    height: number;
    filledCells: Array<{ x: number, y: number }>;
    clickHandler: (x: number, y: number) => void;
}

export const Field: React.FC<FieldProp> = (({width, height, filledCells, clickHandler}) => {
    let divs = [];
    for (let h = 1; h <= height; h++) {
        let cells = [];
        for (let w = 1; w <= width; w++) {
            const isFilled = filledCells.filter(point => point.x === w && point.y === h).length > 0;
            cells.push({
                x: w,
                y: h,
                filled: isFilled
            });
        }
        divs.push(cells);
    }
    return <StyledTable><tbody>
        {divs.map((line, i) => {
            return <StyledTr key={i + 1}>
                {line.map((cell, j) => {
                    return <Cell
                        key={(i + 1) * (j + 1)}
                        x={cell.x}
                        y={cell.y}
                        filled={cell.filled}
                        clickHandler={clickHandler}/>
                })}
            </StyledTr>;
        })}
    </tbody></StyledTable>;
});