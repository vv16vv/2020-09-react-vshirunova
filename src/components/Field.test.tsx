import 'jsdom-global/register';
import React from "react";
import {mount, render} from "enzyme";
import {Field} from "./Field";

describe("Field", () => {

    const clickHandler = jest.fn();

    const filledCellsData = [
        "1, 1",
        "2, 2",
        "3, 4",
    ].map(cellString => {
        const parts = cellString.split(",");
        return {
            x: +parts[0],
            y: +parts[1],
        }
    });

    it("table 1x1 with filled cell", () => {
        expect(render(<Field
            width={1} height={1}
            filledCells={[{x: 1, y: 1}]}
            clickHandler={clickHandler}/>).text()).toEqual("[X] 1, 1")
    });

    it("table 1x1 with empty cell", () => {
        expect(render(<Field
            width={1} height={1}
            filledCells={[]}
            clickHandler={clickHandler}/>).text()).toEqual("[ ] 1, 1")
    });

    it("table 3x4 with different cells", () => {
        expect(render(
            <Field
                width={3}
                height={4}
                filledCells={filledCellsData}
                clickHandler={clickHandler}/>
            ).text()
        ).toEqual("[X] 1, 1[ ] 2, 1[ ] 3, 1[ ] 1, 2[X] 2, 2[ ] 3, 2[ ] 1, 3[ ] 2, 3[ ] 3, 3[ ] 1, 4[ ] 2, 4[X] 3, 4")
    });


    it('table should pass onClick handler to nested cells', () => {
        const x = 1;
        const y = 1;
        const onClickParams = {
            x: x,
            y: y
        }

        const wrapper = mount(
            <Field width={1} height={1} filledCells={filledCellsData} clickHandler={clickHandler}/>
        );

        expect(wrapper).toMatchSnapshot();

        wrapper
            .find('td')
            .simulate('click', {
                target: onClickParams,
            });

        expect(clickHandler).toBeCalledWith(x, y);
    });
});