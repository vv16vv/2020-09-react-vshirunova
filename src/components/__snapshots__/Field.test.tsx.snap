// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field table 1x1 with empty cell 1`] = `
<table
  className="css-12hphec"
>
  <tbody>
    <tr
      className="css-ph6625"
    >
      <td
        className="css-1odp4ej"
        id="id1-1"
        onClick={[Function]}
      >
         
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Field table 1x1 with filled cell 1`] = `
<table
  className="css-12hphec"
>
  <tbody>
    <tr
      className="css-ph6625"
    >
      <td
        className="css-1ws95wt"
        id="id1-1"
        onClick={[Function]}
      >
        X
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Field table 3x4 with different cells 1`] = `
<table
  className="css-12hphec"
>
  <tbody>
    <tr
      className="css-ph6625"
    >
      <td
        className="css-1ws95wt"
        id="id1-1"
        onClick={[Function]}
      >
        X
      </td>
      <td
        className="css-1odp4ej"
        id="id2-1"
        onClick={[Function]}
      >
         
      </td>
      <td
        className="css-1odp4ej"
        id="id3-1"
        onClick={[Function]}
      >
         
      </td>
    </tr>
    <tr
      className="css-ph6625"
    >
      <td
        className="css-1odp4ej"
        id="id1-2"
        onClick={[Function]}
      >
         
      </td>
      <td
        className="css-1ws95wt"
        id="id2-2"
        onClick={[Function]}
      >
        X
      </td>
      <td
        className="css-1odp4ej"
        id="id3-2"
        onClick={[Function]}
      >
         
      </td>
    </tr>
    <tr
      className="css-ph6625"
    >
      <td
        className="css-1odp4ej"
        id="id1-3"
        onClick={[Function]}
      >
         
      </td>
      <td
        className="css-1odp4ej"
        id="id2-3"
        onClick={[Function]}
      >
         
      </td>
      <td
        className="css-1odp4ej"
        id="id3-3"
        onClick={[Function]}
      >
         
      </td>
    </tr>
    <tr
      className="css-ph6625"
    >
      <td
        className="css-1odp4ej"
        id="id1-4"
        onClick={[Function]}
      >
         
      </td>
      <td
        className="css-1odp4ej"
        id="id2-4"
        onClick={[Function]}
      >
         
      </td>
      <td
        className="css-1ws95wt"
        id="id3-4"
        onClick={[Function]}
      >
        X
      </td>
    </tr>
  </tbody>
</table>
`;
